### PR TITLE
Extend String Class

### PR DESCRIPTION
**Pull request by emekanwankwo**

### CODE REVIEW

File: spec/string-spec.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+    it('Should return true if the string contains vowels', () => {
+      expect('home'.hasVowels()).to.be.true;
+      expect('bcd'.hasVowels()).to.be.false;
+    });
+  });
+
+  describe('toUpper', () => {
+    it('Should return the string with all characters in upper case', () => {
+      expect('home'.toUpper()).to.equal('HOME');
+    });
```
Comment (azemoh) [inline spec/string-spec.js:15]: **Testing**
Test for edge cases characters like symbols and numbers.
```javaScript
expect('#$2hhj'.toUpper()).to.equal('#$2HHJ');
```

File: spec/string-spec.js [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+  describe('ucFirst', () => {
+    it('Should return the string with the first character in upper case', () => {
+      expect('home'.ucFirst()).to.equal('Home');
+    });
+  });
+
+  describe('isQuestion', () => {
+    it('Should return true if the string is a question', () => {
+      expect('Are you done?'.isQuestion()).to.be.true;
+      expect('I am done'.isQuestion()).to.be.false;
```
Comment (azemoh) [inline spec/string-spec.js:33]: **Testing**
Test for case where `?` is in the middle of the string.
```javaScript
expect('I am? done'.isQuestion()).to.be.false;
```

File: spec/string-spec.js [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    });
+  });
+
+  describe('words', () => {
+    it('Should return a list of the words in the string as an array', () => {
+      expect('This is a sentence'.words()).to.deep.equal(['This', 'is', 'a', 'sentence']);
+      expect('This is a sentence'.words() instanceof Array).to.be.true;
```
Comment (azemoh) [inline spec/string-spec.js:40]: **Testing**
Also test for edge cases like having unusual multiple spaces.

File: spec/string-spec.js [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+  describe('wordCount', () => {
+    it('Should return the number of words in the string', () => {
+      expect('number of words'.wordCount()).to.equal(3);
+    });
+  });
+
+  describe('toCurrency', () => {
+    it('Should return a currency representation of the string', () => {
+      expect('11111.11'.toCurrency()).to.equal('11,111.11');
```
Comment (azemoh) [inline spec/string-spec.js:52]: **Testing**
Same comment above about testing for edge cases.

File: spec/string-spec.js [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    });
+  });
+
+  describe('fromCurrency', () => {
+    it('Should return a number representation of the currency string', () => {
+      expect('11,111.11'.fromCurrency()).to.equal(11111.11);
```
Comment (azemoh) [inline spec/string-spec.js:58]: **Testing**
Same comment above about testing for edge cases.

File: src/string.js [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+  * // returns true
+  * 'hello'.hasVowels();
+  */
+  hasVowels() {
+    return /[aeiou]/i.test(this);
+  },
+
+  /*
+  * Converts the argument to uppercase characters
+  * @param {string}
```
Comment (azemoh) [inline src/string.js:19]: **JS Doc**
Fix your JS Doc Format.
- First line should have 2 `*` like so `/**`
- `@param` and `@return` identifiers should be capitalised. eg `{String}`

File: src/string.js [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+  * @returns {string}
+  * @example
+  * // returns 'HELLO'
+  * 'hello'.toUpper();
+  */
+  toUpper() {
+    const len = this.length;
+    let cap = '';
+    for (let i = 0; i < len; i += 1) {
```
Comment (azemoh) [inline src/string.js:28]: **Code Quality - Nitpick**
Prefer JavaScript's higher-order functions (`map`, `forEach`) instead of loops like `for-in` for Iterations
As stated in our AirBnB [styleguide](https://github.com/airbnb/javascript#iterators-and-generators)

File: src/string.js [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+  * @param {string}
+  * @returns {string}
+  * @example
+  * // returns 'hello'
+  * 'HELLO'.toLower();
+  */
+  toLower() {
+    const len = this.length;
+    let lower = '';
+    for (let i = 0; i < len; i += 1) {
```
Comment (azemoh) [inline src/string.js:49]: Same comment above about Iterations.

File: src/string.js [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+  /*
+  * Converts the  first character of the argument to uppercase
+  * @param {string}
+  * @returns {string}
+  * @example
+  * // returns 'Hello'
+  * 'hello'.ucFirst();
+  */
+  ucFirst() {
+    return (this[0].toUpper() + this.slice(1));
```
Comment (azemoh) [inline src/string.js:68]: **Code Quality - Critical**
You want to make use of ES6 string interpolation as stated in our style guide. Use template syntax instead of using `+` to concatenate strings. Like so:

``` javaScript
return `${this[0].toUpper()}${this.slice(1)}`;
```

Take a look at our Airbnb [style guide](https://github.com/airbnb/javascript#es6-template-literals) on template syntax.

File: src/string.js [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+  /*
+  * Checks if the argument is a question
+  * @param {string}
+  * @returns {boolean}
+  * @example
+  * // returns true
+  * 'are you coming?'.isQuestion();
+  */
+  isQuestion() {
+    return /[?]$/.test(this);
```
Comment (azemoh) [inline src/string.js:80]: **Code Quality - Nitpick**
No need for range (`[]`) in your ReGex expression as you are just matching one character.
```javaScript
return /?$/.test(this);
```

**Review by azemoh — COMMENTED** [review]
Great use of regular expressions to manipulate strings.
But you need to make sure you cover edge cases when writing tests.

File: spec/string-spec.js (hunk 1 above)
Comment (emekanwankwo) [reply spec/string-spec.js:15]: Thanks for pointing these out, I will work on them.

File: src/string.js (hunk 6 above)
Comment (emekanwankwo) [reply src/string.js:19]: Good eyes to spot those out, I will work on them.

File: src/string.js (hunk 9 above)
Comment (emekanwankwo) [reply src/string.js:68]: Ok, I will work on it.

File: src/string.js [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+  */
+  toLower() {
+    const len = this.length;
+    let lower = '';
+    for (let i = 0; i < len; i += 1) {
+      if (this[i].match(/[A-Z]/)) {
+        lower += String.fromCharCode(this.charCodeAt(i) + 32);
+      } else {
+        lower += this[i];
+      }
```
Comment (elchroy) [inline src/string.js:54]: You might want to consider using tenary operators.
```javascript
lower += this[i].match(/[A-Z]/) ? String.fromCharCode(this.charCodeAt(i) + 32) : this[i];
```
or 

```javascript
lower += this[i].match(/[A-Z]/)
? String.fromCharCode(this.charCodeAt(i) + 32)
: this[i];
```

File: src/string.js [hunk 12]
```diff
@@ -0,0 +1,249 @@
...
+    let alternatingCase = '';
+    for (let i = 0; i < len; i += 1) {
+      if ( (i % 2 !== 0) ) {
+        alternatingCase += this[i].toUpper();
+      } else {
+        alternatingCase += this[i].toLower();
+      }
+    }
+    return alternatingCase;
+  },
```
Comment (elchroy) [inline src/string.js:174]: Consider using a Tenary oeprator here too.

File: src/string.js (hunk 11 above)
Comment (emekanwankwo) [reply src/string.js:54]: Thanks for pointing it out, I have refactored the statements now using tenary operators.
